File: backend/app/schemas/security.py
from __future__ import annotations

from typing import Any, Literal

from app.schemas.base import APIModel
from app.schemas.dashboard import AuditLog


class SecuritySummary(APIModel):
    today_events: int
    blocked_threats: int
    alert_notifications: int
    active_rules: int


class SecurityReportSummary(APIModel):
    total_events: int
    blocked_threats: int
    alert_notifications: int
    active_rules: int
    unique_users: int
    rewrite_events: int
    high_risk_events: int


class SecurityReportBreakdownItem(APIModel):
    key: str
    label: str
    count: int
    share: float


class SecurityReportIncident(AuditLog):
    layer: str | None = None
    verdict: str | None = None
    rule_label: str | None = None
    entity_refs: list["SecurityEntityRef"] = []


class SecurityRule(APIModel):
    id: str
    name: str
    description: str
    type: str
    enabled: bool
    hit_count: int
    last_triggered: str


class SecurityRulesResponse(APIModel):
    summary: SecuritySummary
    items: list[SecurityRule]
    total: int


class UpdateSecurityRuleRequest(APIModel):
    enabled: bool | None = None
    name: str | None = None
    description: str | None = None
    type: Literal["filter", "block", "alert"] | None = None


class SecurityRuleActionResponse(APIModel):
    ok: bool
    message: str
    rule: SecurityRule


class ActiveSecurityPenalty(APIModel):
    user_key: str
    level: str
    detail: str
    status_code: int
    until: str
    updated_at: str


class SecurityPenaltiesResponse(APIModel):
    items: list[ActiveSecurityPenalty]
    total: int


class SecurityPenaltyActionResponse(APIModel):
    ok: bool
    message: str
    user_key: str
    penalty: ActiveSecurityPenalty | None = None
    released_penalty: ActiveSecurityPenalty


class SecurityReportResponse(APIModel):
    generated_at: str
    window_hours: int
    summary: SecurityReportSummary
    status_breakdown: list[SecurityReportBreakdownItem]
    gateway_layer_breakdown: list[SecurityReportBreakdownItem]
    top_resources: list[SecurityReportBreakdownItem]
    top_actions: list[SecurityReportBreakdownItem]
    top_rules: list[SecurityReportBreakdownItem]
    recent_incidents: list[SecurityReportIncident]


class SecurityIncidentReview(APIModel):
    id: str
    timestamp: str
    incident_id: str
    action: Literal["reviewed", "false_positive", "note"]
    note: str
    reviewer: str
    source: str


class SecurityIncidentReviewsResponse(APIModel):
    items: list[SecurityIncidentReview]
    total: int


class CreateSecurityIncidentReviewRequest(APIModel):
    action: Literal["reviewed", "false_positive", "note"]
    note: str = ""


class SecurityIncidentReviewActionResponse(APIModel):
    ok: bool
    message: str
    review: SecurityIncidentReview


class SecurityIncidentReviewListItem(SecurityReportIncident):
    reviewed: bool
    review_status: Literal["pending", "reviewed", "false_positive"]
    last_reviewed_at: str | None = None
    last_reviewer: str | None = None


class SecurityIncidentReviewListResponse(APIModel):
    items: list[SecurityIncidentReviewListItem]
    total: int


class SecurityIncidentReviewActionRequest(APIModel):
    action: Literal["reviewed", "false_positive", "note", "reopen"]
    note: str = ""


class SecurityIncidentReviewActionResult(APIModel):
    ok: bool
    message: str
    incident_id: str
    review: SecurityIncidentReview | None = None


class SecurityEntityRef(APIModel):
    type: Literal["task", "run", "collaboration", "workflow", "user", "channel"]
    id: str
    label: str
    href: str


class SecurityRuleHitDetail(AuditLog):
    layer: str | None = None
    verdict: str | None = None
    rule_label: str | None = None
    entity_refs: list[SecurityEntityRef] = []


class SecurityRuleHitSummary(APIModel):
    total_hits: int
    warning_hits: int
    error_hits: int
    latest_hit_at: str | None = None


class SecurityRuleHitDetailsResponse(APIModel):
    rule: SecurityRule
    summary: SecurityRuleHitSummary
    items: list[SecurityRuleHitDetail]
    total: int


class CreateSecurityRuleRequest(APIModel):
    name: str
    description: str
    type: Literal["filter", "block", "alert"]
    enabled: bool = True


class SecurityRuleVersion(APIModel):
    id: str
    rule_id: str
    timestamp: str
    action: str
    operator: str
    snapshot: SecurityRule
    note: str = ""


class SecurityRuleVersionHistoryResponse(APIModel):
    items: list[SecurityRuleVersion]
    total: int


class RollbackSecurityRuleRequest(APIModel):
    version_id: str


class SecurityPenaltyHistoryItem(APIModel):
    id: str
    timestamp: str
    user_key: str
    action: str
    level: str
    detail: str
    status_code: int
    until: str | None = None
    operator: str
    note: str = ""
    source: str = "audit_log"


class SecurityPenaltyHistoryResponse(APIModel):
    items: list[SecurityPenaltyHistoryItem]
    total: int


class CreateSecurityPenaltyRequest(APIModel):
    user_key: str
    level: Literal["cooldown", "ban"]
    detail: str
    duration_seconds: int
    status_code: int = 429
    note: str = ""
    approval_id: str | None = None
    approval_reason: str | None = None
    approval_note: str | None = None


class ReleaseSecurityPenaltyRequest(APIModel):
    approval_id: str | None = None
    approval_reason: str | None = None
    approval_note: str | None = None


class SecurityRiskProfileItem(APIModel):
    key: str
    label: str
    event_count: int
    blocked_count: int
    warning_count: int
    review_pending: int
    false_positive_count: int
    latest_event_at: str | None = None
    risk_score: int
    entity_refs: list[SecurityEntityRef] = []


class SecurityRiskProfilesResponse(APIModel):
    items: list[SecurityRiskProfileItem]
    total: int


class SecurityTrendPoint(APIModel):
    bucket: str
    total_events: int
    blocked_events: int
    warning_events: int
    false_positive_events: int
    review_events: int


class SecurityTrendResponse(APIModel):
    points: list[SecurityTrendPoint]
    total: int


class SecurityAlertSubscription(APIModel):
    id: str
    channel: Literal["email", "webhook", "nats"]
    target: str
    enabled: bool
    severity_scope: list[str]
    created_at: str
    updated_at: str


class SecurityAlertSubscriptionsResponse(APIModel):
    items: list[SecurityAlertSubscription]
    total: int


class CreateSecurityAlertSubscriptionRequest(APIModel):
    channel: Literal["email", "webhook", "nats"]
    target: str
    enabled: bool = True
    severity_scope: list[str] = ["warning", "error"]


class UpdateSecurityAlertSubscriptionRequest(APIModel):
    target: str | None = None
    enabled: bool | None = None
    severity_scope: list[str] | None = None


class SecurityAlertSubscriptionActionResponse(APIModel):
    ok: bool
    message: str
    subscription: SecurityAlertSubscription


class SecurityReportExportResponse(APIModel):
    generated_at: str
    period: Literal["daily", "weekly"]
    content_type: Literal["markdown"]
    content: str
    summary: dict[str, Any]

File: reception/types/security.ts
import type { LogType } from '@/types/dashboard'

export type SecurityRuleType = 'filter' | 'block' | 'alert'

export interface SecuritySummary {
  todayEvents: number
  blockedThreats: number
  alertNotifications: number
  activeRules: number
}

export interface SecurityReportSummary {
  totalEvents: number
  blockedThreats: number
  alertNotifications: number
  activeRules: number
  uniqueUsers: number
  rewriteEvents: number
  highRiskEvents: number
}

export interface SecurityReportBreakdownItem {
  key: string
  label: string
  count: number
  share: number
}

export interface SecurityRule {
  id: string
  name: string
  description: string
  type: SecurityRuleType
  enabled: boolean
  hitCount: number
  lastTriggered: string
}

export interface SecurityEntityRef {
  type: 'task' | 'run' | 'workflow' | 'user' | 'channel'
  id: string
  label: string
  href: string
}

export interface UpdateSecurityRuleRequest {
  enabled?: boolean
  name?: string
  description?: string
  type?: SecurityRuleType
}

export interface SecurityRulesResponse {
  summary: SecuritySummary
  items: SecurityRule[]
  total: number
}

export interface SecurityRuleActionResponse {
  ok: boolean
  message: string
  rule: SecurityRule
}

export type SecurityPenaltyLevel = 'cooldown' | 'ban'

export interface SecurityPenalty {
  userKey: string
  level: SecurityPenaltyLevel | string
  detail: string
  statusCode: number
  until: string
  updatedAt?: string
}

export interface SecurityPenaltiesResponse {
  items: SecurityPenalty[]
  total: number
}

export interface SecurityPenaltyActionResponse {
  ok: boolean
  message: string
  userKey: string
  penalty?: SecurityPenalty
  releasedPenalty?: SecurityPenalty
}

export type SecurityIncidentReviewAction = 'reviewed' | 'false_positive' | 'note'

export interface SecurityIncidentReview {
  id: string
  timestamp: string
  incidentId: string
  action: SecurityIncidentReviewAction
  note?: string | null
  reviewer?: string | null
  source?: string | null
}

export interface SecurityIncidentReviewsResponse {
  items: SecurityIncidentReview[]
  total?: number
}

export interface CreateSecurityIncidentReviewRequest {
  action: SecurityIncidentReviewAction
  note?: string
}

export interface SecurityIncidentReviewActionResponse {
  ok: boolean
  message: string
  review?: SecurityIncidentReview
}

export interface SecurityPolicySettings {
  messageRateLimitPerMinute: number
  messageRateLimitCooldownSeconds: number
  messageRateLimitBanThreshold: number
  messageRateLimitBanSeconds: number
  securityIncidentWindowSeconds: number
  promptRuleBlockThreshold: number
  promptClassifierBlockThreshold: number
  promptInjectionEnabled: boolean
  contentRedactionEnabled: boolean
}

export interface SecurityPolicySettingsResponse {
  key: string
  updatedAt: string
  settings: SecurityPolicySettings
}

export interface UpdateSecurityPolicySettingsRequest extends SecurityPolicySettings {}

export interface SecurityReportResponse {
  generatedAt: string
  windowHours: number
  summary: SecurityReportSummary
  statusBreakdown: SecurityReportBreakdownItem[]
  gatewayLayerBreakdown: SecurityReportBreakdownItem[]
  topResources: SecurityReportBreakdownItem[]
  topActions: SecurityReportBreakdownItem[]
  topRules: SecurityReportBreakdownItem[]
  recentIncidents: SecurityReportIncident[]
}

export interface SecurityReportIncident extends AuditLog {
  layer?: string | null
  verdict?: string | null
  ruleLabel?: string | null
  entityRefs?: SecurityEntityRef[]
}

export interface SecurityRuleHitSummary {
  totalHits: number
  warningHits: number
  errorHits: number
  latestHitAt?: string | null
}

export interface SecurityRuleHitDetail extends SecurityReportIncident {
  entityRefs?: SecurityEntityRef[]
}

export interface SecurityRuleHitDetailsResponse {
  rule: SecurityRule
  summary: SecurityRuleHitSummary
  items: SecurityRuleHitDetail[]
  total: number
}

export interface CreateSecurityRuleRequest {
  name: string
  description: string
  type: SecurityRuleType
  enabled?: boolean
}

export interface SecurityRuleVersion {
  id: string
  ruleId: string
  timestamp: string
  action: string
  operator: string
  snapshot: SecurityRule
  note?: string
}

export interface SecurityRuleVersionHistoryResponse {
  items: SecurityRuleVersion[]
  total: number
}

export interface RollbackSecurityRuleRequest {
  versionId: string
}

export interface CreateSecurityPenaltyRequest {
  userKey: string
  level: SecurityPenaltyLevel
  detail: string
  durationSeconds: number
  statusCode?: number
  note?: string
}

export interface SecurityPenaltyHistoryItem {
  id: string
  timestamp: string
  userKey: string
  action: string
  level: string
  detail: string
  statusCode: number
  until?: string | null
  operator: string
  note?: string
  source: string
}

export interface SecurityPenaltyHistoryResponse {
  items: SecurityPenaltyHistoryItem[]
  total: number
}

export interface SecurityRiskProfileItem {
  key: string
  label: string
  eventCount: number
  blockedCount: number
  warningCount: number
  reviewPending: number
  falsePositiveCount: number
  latestEventAt?: string | null
  riskScore: number
  entityRefs?: SecurityEntityRef[]
}

export interface SecurityRiskProfilesResponse {
  items: SecurityRiskProfileItem[]
  total: number
}

export interface SecurityTrendPoint {
  bucket: string
  totalEvents: number
  blockedEvents: number
  warningEvents: number
  falsePositiveEvents: number
  reviewEvents: number
}

export interface SecurityTrendResponse {
  points: SecurityTrendPoint[]
  total: number
}

export interface SecurityAlertSubscription {
  id: string
  channel: 'email' | 'webhook' | 'nats'
  target: string
  enabled: boolean
  severityScope: string[]
  createdAt: string
  updatedAt: string
}

export interface SecurityAlertSubscriptionsResponse {
  items: SecurityAlertSubscription[]
  total: number
}

export interface CreateSecurityAlertSubscriptionRequest {
  channel: 'email' | 'webhook' | 'nats'
  target: string
  enabled?: boolean
  severityScope?: string[]
}

export interface UpdateSecurityAlertSubscriptionRequest {
  target?: string
  enabled?: boolean
  severityScope?: string[]
}

export interface SecurityAlertSubscriptionActionResponse {
  ok: boolean
  message: string
  subscription: SecurityAlertSubscription
}

export interface SecurityReportExportResponse {
  generatedAt: string
  period: 'daily' | 'weekly'
  contentType: 'markdown'
  content: string
  summary: Record<string, unknown>
}

export interface AuditLog {
  id: string
  timestamp: string
  action: string
  user: string
  resource: string
  status: Extract<LogType, 'success' | 'warning' | 'error'>
  ip: string
  details: string
  metadata?: Record<string, unknown> | null
}

export interface AuditLogsResponse {
  items: AuditLog[]
  total: number
  limit: number
  offset: number
  hasMore: boolean
}

export interface AuditLogsQuery {
  search?: string
  status?: AuditLog['status']
  layer?: string
  user?: string
  resource?: string
  limit?: number
  offset?: number
}

export type AlertCenterSeverity = 'info' | 'warning' | 'critical'
export type AlertCenterStatus = 'open' | 'acknowledged' | 'resolved' | 'suppressed'

export interface SecurityAlertCenterItem {
  id: string
  sourceType: string
  sourceId: string
  source: string
  severity: AlertCenterSeverity
  status: AlertCenterStatus
  category: string
  title: string
  message: string
  occurredAt: string
  updatedAt: string
  resource?: string | null
  userKey?: string | null
  traceId?: string | null
  workflowRunId?: string | null
  href?: string | null
  metadata?: Record<string, unknown> | null
}

export interface SecurityAlertCenterQuery {
  search?: string
  status?: AlertCenterStatus | 'all'
  severity?: AlertCenterSeverity | 'all'
  source?: string
  limit?: number
  offset?: number
}

export interface SecurityAlertCenterSummaryBreakdownItem {
  key: string
  count: number
}

export interface SecurityAlertCenterSummary {
  total: number
  open: number
  acknowledged: number
  resolved: number
  suppressed: number
  severityBreakdown: SecurityAlertCenterSummaryBreakdownItem[]
  sourceBreakdown: SecurityAlertCenterSummaryBreakdownItem[]
}

export interface SecurityAlertCenterResponse {
  items: SecurityAlertCenterItem[]
  total: number
  summary: SecurityAlertCenterSummary
}

export interface AlertCenterActionRequest {
  note?: string
  durationMinutes?: number
}

export interface AlertCenterActionResponse {
  ok: boolean
  message: string
  alert: SecurityAlertCenterItem
}
